### ISSUE TITLE
Shrinkwrap roads to terrain

### ISSUE THREAD

**Issue by TheDuckCow** [TheDuckCow/godot-road-generator#144]
Already being asked for by a few people, eventually it would be nice if we have a way to shrinkwrap the road to a terrain.

We need to think this feature out some more, but ideas so-far on requirements:

- Must be able to project onto already set up collission meshes
- Must be optional, likely default off for performance. Could make sense to make this be a per RoadPoint setting (so that you could disable or enable it only for relevant stretches)
  - This could pair with a toggle on the RoadContainer which sets/unsets for all RoadPoints within the container. This is instead of it being a RoadContainer-level setting
- Should have a max distance param, which defines the maximum amount down-wards the road should be projected (so it doesn't try to go into deep ravines/consider bridges). Ideally some smoothing between loop cuts that are floating from those on the ground
- Should have a max "angle" which basically says how steeply the road is allowed to change due to shrinkwrapping. This way a small but very deep crevice only makes a subtle difference, instead of the whole road sinking down for just one loop cut.

**Technical ideation:**

- Projection could work in a number of ways:
  - Per vertex point projection: Will be a bumpy road for sure! And the most expensive to compute. We could probably do some smoothing based on the distance of the gutter profile height, to make it a bit less wobbly
  - 2 or 3-point leveling: basically instead of per vertex, compute the height to the left and right of a given edge loop of the road, and possibly a point in the middle. So cross sections are still mostly smooth, but still roughly conforming to the road. Paired with a high-enough gutter profile, this likely could work for most terrains for expected results (if the terrain is so bumpy this doesn't work, it probably realistically needs to be smoothed anyways)
- Raycasting: will be expensive, might be able to use threads for some of this, maybe creating a lookup dict or something. We should also assume that the raycast will be emitted from somewhere well above

Expecting this would be a v1.0+ (ie Godot 4 only) feature.

**Comment by dev-bre** [comment]
is this going to allow the road to adjust itself to the terrain shape? I see that similar tools in unity and UE allow also the other way round.. the road changes the shape of the terrain below (raising the terrain to fill the space).

**Comment by TokisanGames** [comment]
For Terrain3D you can require the user to enable our full terrain collision in the editor. Or you can just detect if we're installed and query heights directly. https://terrain3d.readthedocs.io/en/latest/docs/integrating.html.

As for using raycasts to place your roads, even running thousands of raycasts a second is nbd. We do up to 3000 raycasts per mouse movement event to [detect where the mouse is](https://github.com/TokisanGames/Terrain3D/blob/46f3632b456f2bfcd674a4a1554449fb05658710/src/terrain_3d.cpp#L824). And you'll do it once in the editor, so even if it takes a few seconds after a button push, it's not a concern.

**Comment by TheDuckCow** [comment]
Thanks for sharing the ideas! For sure @TokisanGames I'll be giving this a try once we wrap up/have a first stable godot 4 version. I definitely have the intention of integrating smoothly with as many terrain editors as is realistically feasible.

@dev-bre yup that is a common feature as well, but it the terrain engine itself needs to provide a way to do this. With at least one terrain dev I spoke with, they have a feature to input curves/paths which are used to define areas to smooth out and level the terrain. So as long as we can feed in the road generator's underlying curves through an appropriate callback, we'd be in good shape.

**Comment by TheDuckCow** [comment]
Hi @renzo-massaro! Following up on your question

> Is there any progression with terrain roads? And specifically with the heights and place a road on the different heights??

Unfortunately I don't have news to share on this front. We're stretched pretty thin (e.g. [this PR](https://github.com/TheDuckCow/godot-road-generator/issues/118#issuecomment-2323439662) took quite a bit longer than expected, and [this PR](https://github.com/TheDuckCow/godot-road-generator/pull/173) is proving difficult)

This feature of shrinkwrapping the road to nearby colliders is an area we'd need some external contributing help to get to it in the immediate term as we are otherwise focussed on the milestone features below which are more necessary before more scaled usage of the Road Generator is possible:

https://github.com/TheDuckCow/godot-road-generator/milestone/10

If you have any interest in contributing though, I can definitely point out to where these changes would be made and some thoughts on the design of the feature! It'd be all about performing raycasts in the body of [this function](https://github.com/TheDuckCow/godot-road-generator/blob/828fadb2db60d8f43f2017ff3e28b9a9ba14870f/addons/road-generator/nodes/road_segment.gd#L799), taking into account the displacement amount of the prior/next loop so we can apply some smoothing / shrink angle limitations. This would be a decent new contributor task since it's pretty isolated from anything else like the UI / input callbacks.

**Comment by FishOfTheNorthStar** [comment]
I'm working on this.  So far it integrates tightly with my non-standard Terrain3D implementation, and should port over fairly cleanly to the actual Terrain3D project when it's ready.

Couple things I'm doing may be off spec, most notably that I'm not working from the collider, it'll be entirely off height lookups in the terrain storage.  Basically a 32 bit float grayscale image.  Also, I just can't imagine any form of this that doesn't need to modify the terrain, perhaps a lot.  So I am, for now, implementing an extra custom step in the Terrain3D loading process that overlays an additional height adjustment image on top of the primary elevation, so the modifications can be done non-destructively to the existing heights, and changed or adjusted later.  This adjustment layer would simulate the extensive grading necessary for actual roads, and may occasionally carve it's way through otherwise unpassable peaks, or fill in sharp dips, or add height to smooth out bumps to happen as gradually as possible.

Ultimately I expect this 'terrain modification layer' to also overlay splat-map overrides, too.  so road edges can fade to gravel, and maybe certain very simple roads could be visualized with no other mesh component, just staining the texturing of the existing terrain and adjusting it's height.

How this would translate to other non-Terrain3D systems is a little up in the air.  Maybe the general process could have an abstraction layer in the middle somewhere so other systems can hook into the relevant phases and get access to the modification maps, and provide height data however they do that.

OK so all of that said, I do see a lot of need for purely physics-collision based shrinkwrapping, too... it's just not in my own use-case scenario.  So one thing occurred to me on that... the GPU particles collider maker... that might be very useful for this because it  is not physics based, it's depth map based and it manages the whole process of pulling a big overhead orthographic depth render and translating that into heights, which is most of the work.  It might be useable for the purpose as is, actually.  But at least it could provide a good starting point for similar stuff.  

Also interesting is that particles can use compute shaders I think, so path solving could perhaps be done on the GPU as a particle system that responds to the elevation map or wherever it physics-detected a collision at, and the particle emitter would move from the start, towards the goal, following certain maximum banking/pitch/yaw adjustment parameters?  Just guessing if that's possible but it might be extremely performant, if so.

**Comment by TheDuckCow** [comment]
Hey @FishOfTheNorthStar! Hope you've been well. Curious to hear how your above efforts have come along. Once https://github.com/TheDuckCow/godot-road-generator/pull/216 is merged and we're able to close to v0.6.0 milestone here, we're finally going to be all in on Godot 4 - meaning, the timing will be hopefully more in line to talk about the tactical approaches here. Curious to hear if you have any updated thoughts!

**Comment by FishOfTheNorthStar** [comment]
Hey bud, nice to hear from you.  I've been thinking about your project
lately, and I want to get back to it.  Things have been a little crazy for
me, last time I checked in I think I mentioned I got a new job.  Well I
can't discuss it much, but I took a bit of a retina injury at work, and
I've been busy seeing specialists about treatment, and lawyers about
financial compensation etc.  But all of that should be behind me fairly
soon, and it'll open my schedule up somewhat.  Great to hear you're going
all-in on Godot 4, I think that'll really help simplify development.

Hope you've been well, I'll be around more soon.

**Comment by TheDuckCow** [comment]
So sorry to hear about that @FishOfTheNorthStar, that sounds awful! Truly wishing you the best and hey, really no pressure from our side over here - if it's in the cards and you want to get back to it for sure it'll be appreciated, but don't sacrifice yourself over it.

**Comment by Slluxx** [comment]
Hey, i just want to say that this feature is something i am really looking forward to. It would open up a lot of possibilities for me and the people i am working with. It would be nice to have everything exportable in gltf/glb too, because it would basically completely eliminate the use of blender or other tools to make terrain with roads (aside the cleanup of vertices etc).

Right now, this addon feels a little cartoonish - which is not a bad thing, just not what we are aiming for. Having proper roads and terrain that work well with each other and feel realistic while also keeping performance in mind (for example deleting vertices under the road/splitting the terrain at the edges) would increase our workflow dramatically.

That said, the most important thing is the export of both systems, since we are just using godot as a tool to create environments that will be used in different software later on.


Anyway, thank you guys for developing free tools!

**Comment by TheDuckCow** [comment]
Sure thing @Slluxx I appreciate the validation of the feature requests, it helps to get that affirmation. To cover you, a summary of the state of things today:

* Exporting meshes for external editing:
  * When [this bug](https://github.com/TheDuckCow/godot-road-generator/issues/207#issuecomment-2755066241) is solved, for which a proposed solution was found, this will be available out of the box.
  * That being said, you already have a workaround as described atop the bug linked there + by following [this guide](https://github.com/TheDuckCow/godot-road-generator/wiki/User-guide:-Custom-road-meshes), so you can already do this today, just not as smoothly as ideally.
* This shrinkwrapping feature as described above
  * Currently we don't have someone working on this, a is good area for external contributions. If any members of your team are willing to contribute to this area, I would be more than happy to spend time scoping and guiding where to get started on it. If nothing else, it could be good to cover your workflow expectations.
* Reverse to the above (but can work in tandem): Roads deforming terrains
  * Work on this is in progress (see [discussion here](https://github.com/TheDuckCow/godot-road-generator/issues/215) and a not-quite-useable [draft PR here](https://github.com/TheDuckCow/godot-road-generator/pull/226))
  * This is going to be up each specific terrain integration and depends on whether they support "deletion", so far the plan for Terrain 3D's integration is to not perform deletion since that would be more expensive as you move roads around in the editor (and, I'm not even sure if they support that concept feature).
* Supporting more realistic materials: Though you didn't mention this, I have a hunch this relates to the feeling of being cartoonish:
  * Not surprising, given [the game](https://www.patreon.com/WheelStealGame) for which this was developed :)
  * That being said, it's technically [already possible](https://github.com/TheDuckCow/godot-road-generator/wiki/Creating-custom-materials), but I'll admit the out-of-the-box material undersells what is possible and it's not a trivial operation to just swap in your own textures, since it requires a specific trimsheet format.
  * I'm planning to prepare at least one CC0, PBR-based material which should help mitigate this
  * I haven't created a task for this yet, but will soon

Let me know if that covers everything you can think of - summarizing more to make sure everything is tracked.

**Comment by Slluxx** [comment]
Thanks for this detailed response! That basically covers everything. I personally think that if the road can modify the terrain up or down (with a user defined fallof size), to close gaps that would otherwise exist, is all we (as in me an my team) need. Shrinkwrapping (in the traditional sense) to uneven terrain will most likely lead to undrivable roads in most occasions.

We are mostly working with an old racing simulator, which means that realistic meshes are basically the only thing we can use. Specifically with this in mind, the types of road we are working with are highways, city streets and country roads.

Given the age and restrictions of the simulator, we can not use any shaders from godot or blender. Textures have to be baked, which makes it is really hard to quickly and dynamically create realistic things like highway on/off ramps.

I know this is none of your concern and should not influence how you work on your tool. I am just trying to explain that none of us are good 3D model or texture artists and we are trying to figure out new ways to make better racetracks more quickly. Blender geometry nodes are okay but they tend to be really really slow and modifications after baking meshes are basically impossible without a lot of overhead.


All that aside, i think it is smart to:

- Seperate shoulder from road and use a seperate material for each
- Being able to create intersections between different road nodes dynamically
    - The way i am imagining this is to select any amount of road end nodes and then generating a mesh that connects these points with options for things like an intersection width (0 being a 90° angle on a 4 way intersection, any number higher creating an arc connecting the sides of roads)
- Being able to stencil roadmarkings - possibly as a mesh that cuts away the road it is on or alternativly ontop with some kind of offset. Preferrably an option that the user can decide.
- The ability to add roadside objects between 2 road nodes (or having a seperate roadside path3D node - probably a lot cleaner)

**Comment by TheDuckCow** [comment]
Hi @Slluxx thanks for sharing your use cases here, I'd say a lot of what you mention is planned and only limited by dev time to implement. To summarize and connect to existing issue id's:

> modify the terrain up or down (with a user defined fallof size),

Indeed this is the plan, to have some falloff. So you should be covered there - though if I can ask, what terrain system are you using, custom made or a third party plugin?

> Being able to create intersections between different road nodes dynamically

Yup, this will be a great feature when we can get there. I'll be transparent in saying it's a ways out unless we can get some outside contribution, the task tracker is over here: https://github.com/TheDuckCow/godot-road-generator/issues/121

That being said, the project does ship with a few hand-modeled intersections. I realize it's not your specialty to further edit or create your own geometry, but perhaps modifying provide blend file meshes already provided could be a decent starting point. The latest release has this already.

> Given the age and restrictions of the simulator, we can not use any shaders from godot or blender. Textures have to be baked,

I would argue the current state of the repository is exactly what you want. The pre-modeled intersections that come with the repository right now use the same material as the rest of the roads, which means they are also easily swapped for other materials/textures. This is not a fancy shader, but rather a simple built-in 3D material where a single texture is entirely driven by UV placement. The RoadSegment geo-generator simply sets up the UVs in the appropriate way to make the lanes and shoulders align to the texture trim sheet as necessary, no different than a baked texture would work. Though it's been discussed, we are not actively working on a more dynamic shader solution that is generating lines of different patterns on the fly, if that's what you were referring to avoid.

> Seperate shoulder from road and use a seperate material for each

Can you clarify your use case? Maybe a visual of what you are attempting to do would help. Generally, I imagine the shoulder still being part of the paved surface of the road and thus being tied to the baked texture trim sheet, as we currently have it. If it's about the transition between the road and the terrain itself, I'd say that's somewhat of another matter (but I'm certainly open to inputs and different approaches).

> Being able to stencil roadmarkings - possibly as a mesh that cuts away the road it is on or alternativly ontop with some kind of offset. Preferrably an option that the user can decide.

For additional road markings like left turn symbols or cross walks, though not yet supported, the working idea is to use Godot's decal system with some intentional layer restrictions. Mesh planes could be an alternate mode as well, though can run into zfighting issues. Are you able to share if Godot's decal solution would work for you? [See here for more](https://docs.godotengine.org/en/stable/tutorials/3d/using_decals.html).

Technically speaking, nothing stops you from already doing this manually today, it's just an idea that the plugin here could provide some out of the box workflow to make it a little more streamlined.

**Comment by Slluxx** [comment]
Unrelated to this (but i dont want to create new tickets for it) i noticed that

- creating new road points dont snap to a collision enabled surface (snapping is a recent godot addition)
- multiple road points can't be hightlighted and edited at the same time
- roadgutter profile should probably be controlled by a 2d curve instead of x/y inputs
- roadpoints cant be inserted between existing points (not entirely sure about that)
- paintlines become stretched with lanewidth and shoulder width

The paintlines are probably the hardest thing to "fix", because of the very simplistic uv nature.

I am again bringing up RoadRunner here because their solution is very elegant. Upon export, they create a mesh purely for roadmarkings and use that to cut the road, then apply the new materials. This results in perfect UVs in exchange for a few more vertices. This would also comepletely remove the need for a complicated trimsheet, as you can simply specify a base asphalt texture and each roadmarking can get its own material or share them with other roadmarkings. It would not just fix the stretched UVs & make the trimsheet irrelevant but it would at the same time allow for the roadmarking stencil feature i mentioned.

![Image](https://github.com/user-attachments/assets/d7ddf053-3180-4f90-b9f3-1d172d43d3b0)

![Image](https://github.com/user-attachments/assets/0c994e9f-7c44-4188-8736-1c87d4229998)

**Comment by Slluxx** [comment]
@TheDuckCow Sorry for the mention but since we had an ongoing convertsation i figured you might have forgotten or overlooked this.

**Comment by TheDuckCow** [comment]
Thanks for flagging @Slluxx, I wanted to give a fuller answer but didn't have the time to sit down - and to also direct this convo into another, since this has strayed from the topic of shrinkwrapping.

If you're ok with it, can you copy your last response or two into #145 as it's pretty related, for the parts related to decals/materials. And then it'll give me a moment to collect my thoughts some more. 

For your other comments:

> creating new road points dont snap to a collision enabled surface (snapping is a recent godot addition)

Placing new roadpoints via left click in the connect mode, with collisions in the scene, should work this way. Maybe I need some more examples of the scenario you are facing?

> multiple road points can't be hightlighted and edited at the same time

This is a new limitation that appeared with Godot 4, it was possible in Godot 3. We'd have to do some research to see if this is possible again. Ticket created here so we don't lose track: #237

> roadgutter profile should probably be controlled by a 2d curve instead of x/y inputs

That's reasonable - I would say this falls into the broader bucket of us wanting to provide more roadside options out of the box aside from just the tapered edge (e.g. a default sidewalk setup would be very convenient). Some overlap with #123 but the curve (and samples loop cuts?) could still be helpful on its own. We'd just need to contemplate how it interpolates nicely between RoadPoints.

> roadpoints cant be inserted between existing points (not entirely sure about that)

Agreed that having a subdivide or option to draw in new points would be helpful. I've created #236 for now, though it's a lower priority since (per the description/demo on that ticket) there's a workaround that exists today which just takes a couple more clicks.

> road meshes should optionally be split at 65,000 vertices to ensure compatibility with software that is limited to 16-bit indexing

Would be resolved with some level of implementation of #233, the part where it breaks up longer road segments.

> paintlines become stretched with increased lane- or shoulder-width

Aligns with discussion on #145 I'd say.

**Comment by TheDuckCow** [comment]
Copying another +1 for this feature to this thread, in @wikicode96's words from https://github.com/TheDuckCow/godot-road-generator/issues/403:

---

I think the workflow would be much more intuitive if the **road adapted to the terrain instead of the terrain adapting to the road**.

The idea is similar to how road placement works in **Cities: Skylines**:

1. Create and sculpt the Terrain3D terrain.
2. Create hills, mountains, valleys, etc.
3. Add a road on top of the terrain.
4. The road automatically follows the terrain height.
5. If the terrain is later modified or sculpted, the road automatically updates its height to remain attached to the terrain.

In other words, the terrain would be the **source of height information**, and the road would conform to it.

The main idea is not necessarily to change the road geometry itself, but to make the road's vertical position follow the Terrain3D surface, while Terrain3D handles the appropriate blending around the road.

Would something like this be possible with the current Terrain3D integration? I'd like to research this feature.

**Comment by TokisanGames** [comment]
Probably wiser and a better user experience. Look at https://github.com/TokisanGames/Terrain3D/blob/main/project/addons/terrain_3d/utils/terrain_3d_objects.gd which already adjusts mesh objects in realtime when the terrain is sculpted.

**Comment by TheDuckCow** [comment]
Nice, thanks for the sample script, that could definitely be interesting. In my mind, the two features in general both have their merits in the end, but they are different use cases/workflows. In some games, you end up needing to really nail down the road layout first, and derive more of the terrain afterwords. In other cases yes, it's terrain first all the way. 

What's interesting about this and was slightly different from how I was initially considering this, is thinking of actually moving the RoadPoint controls, with the terrain, rather than just having each segment have a distance of influence. So in the end, there could be really like 3 levels of influence:

- 1. Local terrain flattening (what exists today: Road gen modifies terrain)
- 2. Road shrinkwrapping to terrain (original idea of this ticket, project the road down onto terrain so it conforms)
- 3. Control RoadPoint snapping - the new idea here, which could be blended with _either_ the shrinkwrapping (e.g. more like dirt or bumpy road) or terrain flattening (paved road/highway with some gutter).

While number 2 would be nice, point 3 also seems like a no brainer to implement if we can get the right callbacks. I'd consider even implementing this in a more generic way that goes beyond the terrain 3D integration, though maybe that's more like a separate connector type e.g. a collider group. To get point 3 to work with point 1 at the same time might require some timer trickery/safety checks, but if there's a will there's a way. Probably could define since code for a dead zone so if the terrain height directly under a given roadpoint changed by less than some amount, it doesn't trigger any change (so that the flattening doesn't trigger a recursive action).

**Comment by wikicode96** [comment]
I was planning to make the changes to my game locally, since I need a more comfortable workflow. Once I have it ready, I could share the code with you, and if you like it, we can merge it.

**Comment by TheDuckCow** [comment]
Sure @wikicode96, if you are already in progress or have a clear idea what you want to do, I could definitely review what you create. That being said, I have some specific thoughts  for how this should work in practice. My initial ideas here, to consider guiding your approach if you're open to it (and thus, making it more likely it could be merged in or at least incorporated in part):

Overview: Shrinkwrapping roads down to a terrain should be a **general purpose** feature of the road generator, and not specific to the terrain 3D connector. There should be a common hook / mechanism to push terrain changes to the Road generator so it can respond accordingly, somewhat independent of its source. The `RoadTerrain3DConnector` can remain the intermediate glue to pipe signals/data calls as needed between the two, thus remaining generalized such that an equivalent integration could be made with any other terrain tool or even just raw collider meshes.

The notes below are a proposal of final state functionality, but definitely could be broken into simpler initial working states (e.g. those smoothing / preprocessing steps mentioned down below could be out of scope).

1. I'm expecting the best option is to define a "terrain" group property on the RoadManager, which individual nodes would be added to
   - When the `RoadTerrain3DConnector` is added, it would auto register itself to that group
3. Then on Road Manager setup, it would query over the group:
   - Generate a lookup of bounding boxes for each object, so we can more efficiently know which registered nodes to use for flattening at any given time. For terrain 3D use cases, it'd be just that most often; but someone using a series of manually defined static trimesh bodies or even a combination of different primitive shapes, they would also all be added
   - Hook according signals from those source nodes to the according terrain flattening function interface on the Road Manager. That would be minimally the signals corresponding to move/scale/rotate
   - For Terrain3D it'd be the `terrain.data.maps_edited.connect` signal (will need to look at that sample script to ensure we can do this in a sparse way)
     - Key challenge here would be how to avoid a "deep" connection inside the Road Generator itself, since we don't want to have the core Road Generator addon reference Terrain3D types (that's what the connector is for, and allows for more types of connectors in the future).
     - Instead, the `RoadTerrain3DConnector` should be the one to actually handle the terrain 3D signal, and bubble that into a generalized format to pass along to the Road Generator, perhaps a grid array of height points or just a reference to the collider + relevant bounding box.
     - Unsure if the RoadTerrain3DConnector will need to define a pass through function for getting the height of a given point, or just a way of sending along the actual collider. I also hear what TokisanGames has said about directly querying the height data, so perhaps there's a way to do that too and not even depend on editor collisions being enabled.
4. Then determine when a "relevant change" is made:
   - When any of the aforementioned signals are emitted
   - Any internal road modification (moving a RoadPoint, adding/removing points, changing configurations or lane widths, all those usual things)
5. When that change is triggered, the RoadManager responds to the event
   - If it was a terrain node change which triggered it, determine the relevant bounding box of changes to then determine which RoadSegments and RoadIntersections are relevant to be rebuilt.
   - If it was a road generator initiated change, then we just rebuild the segments that would have been rebuilt anyways
   - In the RoadIntersection (if feasibly scoped) and RoadSegment classes (segments between two points), we would then apply the shrinkwrapping steps there by performing raycasts downwards, scoped to only the nodes in the relevant group, possibly reduced to just those with relevant bounding boxes relating to our RoadSegment.
6. Finally, apply the shrinkwrapping while generating the road segments:
   - There likely needs to be a prepass that calculates the raw raycast heights all the way along a given segment curve (with the same density as loop cuts to be) to determine the amount of smoothing necessary for each loop cut. We could either just do a single raycast at the center of the road, or attempt to support some amount of tilting by raycasting points from the left and right edges. Main purpose of this prepass is to then bring into a smoothing function to see if any points need to be "lifted" and treated as floating
     - This prepass is what would allow us to gracefully handling things like sharp ravines or cliffs, or transitioning into bridge segments. Perhaps some users will want to define a max incline rate, so the Road Generator (as much as possible) tries to solve for a layout to match this figure.
     - This prepass would also have to intelligently account for the edges of a given segment, so there's not discontinuities. It probably means to calculate final prepass of one segment, it has to calculated the raw values a little bit into each of the connected segments, depending on the "range" of the smoothing property in meters.
   - Then, a second pass does the actual geo generation, and this time (based on a conformity range 0-1 plus the target float amount value) could do the raycasts per vertex from the top of the road.
   - Conformity 1.0 = exactly match down to the ground below (offset by half the gutter y value, to avoid clipping), will feel more like a bumpy dirt road to drive over
   - Confirmity 0.0 = no raycast per individual vert, instead placement is just based off the smoothed prepass lines. Road will feel smooth to drive over;
   - Conformity 0.5 = a linear blend between the two.
7. Shrinkwrapping for itnersections:
   - The first pass is just seeing where the RoadPoints end up once vertically displaced, as well as the intersection center. If we do define a max % grade property, this could also enforce a maximum level of height differential between the edge RoadPoints.
   - "Second pass" has the same conformity value doing raycasts per vert
8. Finally, users may still choose to run the terrain flattening step as a post step here, to avoid any terrain poking through the middle of the roads. This would have some challenges though, as the destructive editing nature of flattening means a) creating an min influence buffer to avoid a recursive road flattening/shrinkwrapping loop, but also it means that an initially partially bump road gets totally smoothed out the next time. So, would require some thought.
- Unknown:
  - is there a way to get a signal or react to new nodes being added to a group? if so, we could automatically perform updates this way.
  - Will we need to introduce some signal processing delays, so while you are actively editing terrain, it's not constantly at 60fps trying to regenerate roads? Perhaps it's a setting on the connector or the road generator itself. Could be this time is used to wait until there's been e.g. 200ms of no terrain edits before recalculating anything (so a long continuous edit wont' see any road changes until you stop), or limit updates to occur at most every 200ms or 1s (so you'll see updates as you continuously edit the terrain, but spaced out). If we have the post process flattening step as an allowed use case, I imagine we'd need to wait for a single edit to actually complete.
